File: python/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        buy, sell = 0, 1 
        
        while sell < len(prices):
            if prices[buy] < prices[sell]: # is it profitable?
                profit = prices[sell] - prices[buy]
                max_profit = max(profit,max_profit)
            else:
                buy = sell #we move our buy pointer wherever "sell" is, as it is lower and we want to buy at a lower price
            sell += 1 #and we advance the sell pointer
        
        return max_profit

File: javascript/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let max_profit = 0;
    let min = prices[0];
    
    for(let i = 1; i < prices.length; ++i) { 
        if(min > prices[i]) {
            min = prices[i];
        } else if(prices[i] - min > max_profit) {
            max_profit = prices[i] - min;
        }
    }
    
    return max_profit;
};
